client/state/reader/posts/reducer.js

```diff
@@ -20,7 +20,19 @@ export function items( state = {}, action ) {
 	switch ( action.type ) {
 		case READER_POSTS_RECEIVE:
 			const posts = action.posts || action.payload.posts;
-			return { ...state, ...keyBy( posts, 'global_ID' ) };
+			const postsByDedupedGlobalId = {};
+
+			// Assign a new key in case of global_ID collision.
+			for ( const [ globalId, post ] of Object.entries( keyBy( posts, 'global_ID' ) ) ) {
+				let key = globalId;
+				if ( state[ globalId ] && state[ globalId ].feed_item_ID !== post.feed_item_ID ) {
+					key = `${ globalId }-${ post.feed_item_ID }`;
```

Comment:
Instead of looping to find a collision and assigning a different key if there is one,  would it make sense to just change the original key to something like this in the first place?